Added FPGA device_global tutorial (#1)
artemrad/oneAPI-samples: DirectProgramming/DPC++FPGA/Tutorials/Features/device_global/README.md

### How to initialize a `device_global` instance
A `device_global` instance is always zero-initialized. If there is a known first usage of a `device_global` in device code, there exists a technique, described below, that allows a `device_global` to be initialized to a non-zero value. 

Instantiate a second `device_global<bool>` representing a flag that will control when to initialize the `device_global` to a non-zero value. This flag will get zero-initialized to `false`. Once initialization happens, the flag is set to `true` and initialization code doesn't execute on subsequent relaunches of the kernel.

```cpp
using FPGAProperties = decltype(sycl::ext::oneapi::experimental::properties(
    sycl::ext::oneapi::experimental::device_image_scope));

sycl::ext::oneapi::experimental::device_global<int, FPGAProperties> val;
sycl::ext::oneapi::experimental::device_global<bool, FPGAProperties> is_val_initialized;

int main () {
  sycl::queue q;
  q.submit([&](sycl::handler& h) {
    h.single_task([=] {
      // Initialization happens only once
      if (!is_val_initialized) {
        // do initialization

>         is_val_initialized.get() = true;

Reviewer: ```
// Set an initial value
val.get() = 42;
```

# `device_global`

> This FPGA tutorial explains how to use `device_global` class as a way of keeping state between multiple kernel invocations.

Reviewer: The first time I read this sentence I thought it meant to keep a state across multiple kernels, what do you think of:
```suggestion
This FPGA tutorial explains how to use `device_global` class as a way of keeping a state between multiple invocations of a kernel.
```

| Optimized for                     | Description
|:---                                 |:---
| OS                                | Linux* Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA; <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel&reg; FPGA 3rd party / custom platforms with oneAPI support <br> **Note**: Intel&reg; FPGA PAC hardware is only compatible with Ubuntu 18.04*
| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit

> | What you will learn               | The basic usage of the `device_global` class <br> How initialize a `device_global` to non-zero values<br>

Reviewer: ```suggestion
| What you will learn               | The basic usage of the `device_global` class <br> How to initialize a `device_global` to non-zero values, and in what cases this is useful<br>
```

Author: I agree with the `to` article suggestion. But i don't think i cover what cases non-zero initialization is useful for

Reviewer: Yep, that is what I expected to see in the sample as I did not understand what you meant by the non-zero initialization at the time - I think that you can remove the second sentence entirely. You are basically showing what a `device_global` is and how to use it

| Time to complete                  | 15 minutes

## Purpose
This tutorial demonstrates a simple example of initializing a `device_global` to a non-zero value, and using it to keep state between multiple re-launches of a kernel.

### Description of `device_global`

> `device_globals` are an extension that introduces device scoped memory allocations into SYCL that can be accessed within a kernel using syntax similar to C++ global variables, but that have unique instances per `sycl::device`. Similar to C++ global variables, `device_globals` have namespace scope and are visible to all kernels within that scope.

Reviewer: device_global**s** is no longer a SYCL keyword 
What are your thoughts on:
```suggestion
The `device_global` class is an extension that introduces device scoped memory allocations into SYCL that can be accessed within a kernel using syntax similar to C++ global variables, but that have unique instances per `sycl::device`. Similar to C++ global variables, a `device_global` variable have a namespace scope and is visible to all kernels within that scope.
```

> `device_globals` can be templated on a list of compile-time-constant `properties` that can enable various optimization behaviors. One such `property` is `device_image_scope` which is used in the code sample. It limits the scope of a single instance of a `device_global` to a `device_image`, further details on this property can be found [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc#properties-for-device-global-variables). The absence of the `device_image_scope` property is currently not supported by the compiler.

Reviewer: Similarly to the last comment:
```suggestion
A `device_global` instance can be templated on a list of compile-time-constant `properties` that can enable various optimization behaviors. One such `property` is `device_image_scope` which is used in the code sample. It limits the scope of a single instance of a `device_global` to a `device_image`, further details on this property can be found [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc#properties-for-device-global-variables). The absence of the `device_image_scope` property is currently not supported by the compiler.
```

If `device_image_scope` is required for a `device_global`, maybe it would be worth going into a bit more details here.
Something like "Each `device_global` variable must be associated to a scope, and contrary to its name is not completely _global_ to the entire SYCL program. Instead it is global to a device. The `device_image_scope` property allows to set this scope to the `device_image` scope which corresponds to the entire FPGA." 

If I'm wrong at this point, it is because I understood incorrectly how this work and we need to add more explanations :p

Author: I was really avoiding of going into the details of `device_image_scope` in the tutorial. The details are complicated and the wording that is the spec has been worked on for at least a couple of months by Intel's programming language working group. 

I think it is important to mention just because it is used in the tutorial. 

I am all ears for any suggestion on how to achieve this. 

Your understanding is not wrong but also not right, it is over simplified. Vanilla `device_globals` have a device scope. With the `device_image_scope` property the scope changes to a `device_image`. The problem is the `device_images` are not designed to be easily manipulated by the SYCL API users, and how many `device_images` are created for a program is target defined. More over there is no programmatic way of determining when device_images have been swapped - hence the official wording in the spec is quite long in order to convey the logic that the user needs to go into to reason about when `device_images` gets swapped in and out. 

For the FPGA target, in most cases (!) we should have a single device_image which simplifies away a lot of the complicated reasoning away.  But we still have a hard time communicating with the user when the act of "programming the FPGA" happens, since this is an implementation detail and has nuance.

> `device_globals` can be used to store state across multiple re-launches of a kernel without having to pass in a `buffer` as a kernel argument. This is useful in cases where components are nodes in a state-machine.

Reviewer: ```suggestion
A `device_global` instance can be used to store state across multiple re-launches of a kernel without having to pass in a `buffer` as a kernel argument. An example of an application that would benefit from such a state is where kernels are nodes in a state-machine.
```

> ### Non-zero-initialization of a `device_global` - pre C++20

Reviewer: Just looking at the title, this sounds premature: As a user I would have liked to have a quick example of the interface of a `device_global` and possibly a toy example
Now this sounds like we are diving in a corner case without me having seen the general case

Author: Any suggestions for a title change? This is a toy example, as you have seen. 

I think it is important to cover the initialization technique for our users - but i don't think it is big enough to warrant a second tutorial. If you disagree - i am open to discuss this.

Reviewer: I think you should just remove the mentions to "Non-zero-initialization" everywhere and just have a "How to use the  device_global class"
Plus, it's not really "non-zero" right? it's more like "non-default-value" (I'm thinking about booleans here)
I feel like it simplifies things a lot if you just show the use model by initializing a device global ,as most users should do I guess ( just to avoid not knowing if the value is initialized or not by default, and is this going to be the same value in emulation vs in HW, etc. the typical non initialized value questions)

> Without C++20 `consteval` keyword support, `device_globals` are zero-initialized. If there is a known first usage of a `device_global` in device code, there exists a technique, described below, that allows a `device_global` to be initialized to a non-zero value. 

Reviewer: ```suggestion
Without C++20 `consteval` keyword support, a `device_global` is always zero-initialized. If there is a known first usage of a `device_global` in device code, there exists a technique, described below, that allows a `device_global` to be initialized to a non-zero value. 
```

I did not understand this paragraph, I'm guessing that this is because I don't know what the initialization value means. So this tells me there is a way to have a non-zero initialization value, but at this point I don't know what a 0 value would do and what is the motivation for using a non 0 value

Author: My intent was to rely on the C++ terminology of [zero-initialization](https://en.cppreference.com/w/cpp/language/zero_initialization) and [value-initialization](https://en.cppreference.com/w/cpp/language/value_initialization]) so that it is intuitive to the reader what is the behavior/benefits of each. 

What will help in order to clarify the situation. Again i want to avoid truing to define these terms here because they are complicated and have nuances depending on the type that is being initialized.

Reviewer: Yeah, I would personally just says "note that the technique below for initializing the `device_global` variables will be simplified once we support C++20 to `device_global<int> a {42};`"

> Using a second `device_global<bool>` representing a flag that will control when to initialize the `device_global` to a non-zero value. This flag will get zero-initialized to `false`. Once initialization happens, the flag is set to `true` and initialization code doesn't execute on subsequent relaunches of the kernel.

Reviewer: Similarly, I did not understand this paragraph because I'm missing the context

```cpp
using namespace sycl;

> using namespace sycl::ext::oneapi::experimental;

Reviewer: Please don't use `using namespace` and use the explicit namespaces so that users know where each thing lives.
If this feature is experimental, you'll need to move it to `DirectProgramming/DPC++FPGA/Tutorials/Features/Experimental`

Author: Will move all of this to the Experimental folder as a last action item - im afraid all of the comments you gave me will get lost if i do this now.

device_global<int> val;
device_global<bool> is_val_initialized;

int main () {

>   queue Q;

Reviewer: `Q` is a variable so should be snake_case `q`

  Q.submit([&](sycl::handler& h) {
    h.single_task([=] {
      // Initialization happens only once

>       if (!is_val_initialized) {

Reviewer: What happens with concurrent accesses? can race-conditions occur? is there a semaphore/mutex implementation available for this use case? or is this usage restricted to the same kernel being called over and over again so we know there is only one point of access? what happens to this global variable if it is in a parallel region of a single task doing e.g. a reduction?
(basically all the usual race-condition concerns :) )

Author: Right now we do not support accessing a specific `device_global` from multiple kernels. So race conditions of that sort are impossible. 

I guess the user might stick this into a pipelined loop and possible get a race condition.... on second thought i don't think it is possible, since there will be a write_after_read dependency and we will not start the next iteration of the loop in this code block until the dependency is resolved.

        // do initialization
        is_val_initialized.get() = true;
      }

      // uses of `val`
    });
  });
}
```

> 

Reviewer: Note that at this point I still don't know what this non-zero-initialization values means :)

Author: That is not good. It means that teh device_global has been initialized to a value that is non-zero prior to any uses. Initialization by definition only should happen once, in other words subsequent launches of the kernel should not re-initialize the device_global

Reviewer: Honestly, as I mentioned after I was just overthinking what the non-zero-initialization was. 
I was thinking that it had a HW effect - such as dimensioning a pipe.
In the end it's just the first time use of the device_global and is simply the regular use model
That's why I'm suggesting to remove the titles mentioning the non-zero-initialization as it makes it appear as a complex corner case while it's simply the default use model

> ### Non-zero-initialization of a `device_global`- with C++20

Reviewer: Should we omit the pre/post-C++20 discussion and just update the tutorial when C++20 is supported? 
I'm not sure this brings much value to the users understanding of how to use a `device_global`

Author: Yeah i was wrestling with that when i was writing this. Ultimately i decided that it would be really natural for user to want constant value-initialization on device_globals - and i wanted to preemptively stop feature requests and explain that it is in the works and is gated by a modern C++20 feature that is not fully implemented by any compiler.

Do you still think i should cut this?

Reviewer: I think that in that case you should just have a quick mention as a note in a "Use model" section saying something like what I wrote above: 
"note that the technique below for initializing the device_global variables will be simplified once we support C++20 to device_global<int> a {42};"

> In a future version of this extension, it is expected that when C++20 support is available and enabled, the `consteval` keyword will be used to enable compile-time constant initialization of the device allocations backing device_global. This will simplify some coding patterns, compared with the current zero-initialization requirement.

Reviewer: ```suggestion
In a future version of this extension, it is expected that when C++20 support is available and enabled, the `consteval` keyword will be used to enable compile-time constant initialization of the device allocations backing `device_global`. This will simplify some coding patterns, compared with the current zero-initialization requirement.
```

```cpp
using namespace sycl::ext::oneapi::experimental;

> device_global<int> val {42};

Reviewer: I get how you'll be able to initialize the `device_global` with `{}` in the future but the paragraph above goes into too much implementation details (knowing that `consteval` is what is missing is not useful to the end user).

 A simple "this is how we'll do it when C++20 is supported" is enough (if you decide to keep the C++20 discussion, which I would suggest you remove)

```

### Additional Documentation
- [SYCL `device_global` Language Specification](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc) helps understand the API and of `device_globals` and restrictions on it's usage
- [Explore SYCL* Through Intel&reg; FPGA Code Samples](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of FPGAs and SYCL.
- [FPGA Optimization Guide for Intel&reg; oneAPI Toolkits](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) helps you understand how to target FPGAs using SYCL and Intel&reg; oneAPI Toolkits.
- [Intel&reg; oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) helps you understand target-independent, SYCL-compliant programming using Intel&reg; oneAPI Toolkits.

## Key Concepts
* The basic usage of the `device_global` class
* How initialize a `device_global` to non-zero value

> 

Reviewer: At this point I completely understood what you meant by "non-zero-initialization" and I think that as it is presented now it makes the user believe that this is a much more complex thing that is going to be presented (such as a value passed to a template parameters that changes the where the data lives and so on).
So my suggestion would be to remove everything that says "non-zero-initialization" and just present it as "how to initialize a `device_global` variable". Then you can have a simple note if you want saying "once we support C++20 this will be reduced to <...>".

Now I'm thinking again about something you mentioned at the beginning: the `device_image_scope` property that does not appear in this example + what not explained in more details.

Another remark is that you never mentioned what are the template parameters of `device_global`  class. So I was confused at first with your `device_global<bool>` but I then understood with the code example that the template is `device_global<type> ...` just like one would use "type ..." to declare a variable at the global scope in C.

Author: Great suggestions for paragraph 1 and 2. 

Good catch with paragraph 3, ill address that.

> ## Using Visual Studio Code*  (Optional)

Reviewer: Please move this section after "Running Samples in Intel DevCloud"

Author: Huh? I copied this file from another code sample - so that code sample has the exact same problem. I think the tutorial is `private_copies`

Author: Do you have an example code sample that has them in teh right order?

Reviewer: Hum interesting, it seems that the order is not the same in all samples :/ 
Don't worry about it, I'm trying the convince the code sample committee to move these repeated sections to a top level readme, so this would end up being remove hopefully :)

If an error occurs, you can get more details by running `make` with
the `VERBOSE=1` argument:
``make VERBOSE=1``
For more comprehensive troubleshooting, use the Diagnostics Utility for
Intel&reg; oneAPI Toolkits, which provides system checks to find missing
dependencies and permissions errors.
[Learn more](https://www.intel.com/content/www/us/en/develop/documentation/diagnostic-utility-user-guide/top.html).

### In Third-Party Integrated Development Environments (IDEs)

You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [FPGA Workflows on Third-Party IDEs for Intel&reg; oneAPI Toolkits](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html).

## Running the Sample

 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
     ```
     ./device_global.fpga_emu     (Linux)
     device_global.fpga_emu.exe   (Windows)
     ```
2. Run the sample on the FPGA device:
     ```
     ./device_global.fpga         (Linux)
     device_global.fpga.exe       (Windows)
     ```

### Example of Output
```

> PASSED_fpga_compile

Reviewer: From just looking at this message, I'm not sure what it's suppose to say.
If we are running the exe, surely the compile passed?

Author: Should i print out intermediate results? This simple example does not need any timing output. It simply just makes sure that the value of the counter is what is expected.
